File: backend/src/models/Member.js
'use strict';

const { Model, DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  class Member extends Model {
    static associate(models) {
      // Define associations here
      Member.hasMany(models.Dependent, {
        foreignKey: 'member_id',
        as: 'dependents'
      });

      Member.hasMany(models.Transaction, {
        foreignKey: 'member_id',
        as: 'transactions'
      });

      Member.hasMany(models.Transaction, {
        foreignKey: 'collected_by',
        as: 'collected_transactions'
      });

      // Self-referencing association for family relationships
      Member.belongsTo(Member, {
        foreignKey: 'family_id',
        as: 'family_head'
      });

      Member.hasMany(Member, {
        foreignKey: 'family_id',
        as: 'family_members'
      });

      // Outreach notes
      if (models.Outreach) {
        Member.hasMany(models.Outreach, {
          foreignKey: 'member_id',
          as: 'outreach_notes'
        });
      }

      // Tasks assigned to this member
      if (models.DepartmentTask) {
        Member.hasMany(models.DepartmentTask, {
          foreignKey: 'assigned_to',
          as: 'assigned_tasks'
        });
      }
    }
  }

  Member.init({
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true,
      allowNull: false
    },
    firebase_uid: {
      type: DataTypes.STRING(255),
      allowNull: true,
      unique: true,
      field: 'firebase_uid'
    },
    first_name: {
      type: DataTypes.STRING(100),
      allowNull: false,
      validate: {
        notEmpty: true
      },
      field: 'first_name'
    },
    middle_name: {
      type: DataTypes.STRING(100),
      allowNull: true,
      field: 'middle_name'
    },
    last_name: {
      type: DataTypes.STRING(100),
      allowNull: false,
      validate: {
        notEmpty: true
      },
      field: 'last_name'
    },
    email: {
      type: DataTypes.STRING(255),
      allowNull: true,
      unique: true,
      validate: {
        isEmail: true
      }
    },
    phone_number: {
      type: DataTypes.STRING(20),
      allowNull: false,
      unique: true,
      validate: {
        notEmpty: true
      },
      field: 'phone_number'
    },
    date_of_birth: {
      type: DataTypes.DATEONLY,
      allowNull: true,
      field: 'date_of_birth'
    },
    gender: {
      type: DataTypes.ENUM('male', 'female', 'other'),
      allowNull: true
    },
    marital_status: {
      type: DataTypes.ENUM('single', 'married', 'divorced', 'widowed'),
      allowNull: true,
      field: 'marital_status'
    },
    baptism_name: {
      type: DataTypes.STRING(100),
      allowNull: true,
      comment: 'Baptism name',
      field: 'baptism_name'
    },
    repentance_father: {
      type: DataTypes.STRING(100),
      allowNull: true,
      comment: 'Name of the repentance father',
      field: 'repentance_father'
    },
    household_size: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 1,
      comment: 'Number of people in the household',
      field: 'household_size'
    },
    street_line1: {
      type: DataTypes.STRING(200),
      allowNull: true,
      field: 'street_line1'
    },
    apartment_no: {
      type: DataTypes.STRING(50),
      allowNull: true,
      field: 'apartment_no'
    },
    city: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    state: {
      type: DataTypes.STRING(50),
      allowNull: true
    },
    postal_code: {
      type: DataTypes.STRING(20),
      allowNull: true,
      field: 'postal_code'
    },
    country: {
      type: DataTypes.STRING(100),
      allowNull: true,
      defaultValue: 'USA'
    },
    emergency_contact_name: {
      type: DataTypes.STRING(200),
      allowNull: true,
      field: 'emergency_contact_name'
    },
    emergency_contact_phone: {
      type: DataTypes.STRING(20),
      allowNull: true,
      field: 'emergency_contact_phone'
    },
    yearly_pledge: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Yearly membership pledge amount in USD',
      field: 'yearly_pledge'
    },
    date_joined_parish: {
      type: DataTypes.DATEONLY,
      allowNull: true,
      field: 'date_joined_parish'
    },
    spouse_name: {
      type: DataTypes.STRING(200),
      allowNull: true,
      field: 'spouse_name'
    },
    family_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      references: {
        model: 'members',
        key: 'id'
      },
      onUpdate: 'CASCADE',
      onDelete: 'SET NULL',
      field: 'family_id'
    },
    role: {
      type: DataTypes.ENUM('member', 'admin', 'treasurer', 'secretary', 'church_leadership', 'relationship', 'guest', 'deacon', 'priest'),
      allowNull: false,
      defaultValue: 'member'
    },
    is_active: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
      field: 'is_active'
    },
    is_imported: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
      field: 'is_imported'
    },
    registration_status: {
      type: DataTypes.ENUM('pending', 'complete', 'incomplete'),
      allowNull: false,
      defaultValue: 'pending',
      field: 'registration_status'
    },
    // Onboarding / Welcome tracking
    is_welcomed: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
      comment: 'Whether the member has been welcomed by the Relationship Department',
      field: 'is_welcomed'
    },
    welcomed_at: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Timestamp when the member was welcomed',
      field: 'welcomed_at'
    },
    welcomed_by: {
      type: DataTypes.BIGINT,
      allowNull: true,
      comment: 'Member ID of the staff/admin who marked as welcomed',
      references: {
        model: 'members',
        key: 'id'
      },
      onUpdate: 'SET NULL',
      onDelete: 'SET NULL',
      field: 'welcomed_by'
    },
    interested_in_serving: {
      type: DataTypes.ENUM('yes', 'no', 'maybe'),
      allowNull: true,
      defaultValue: 'maybe',
      comment: 'Whether the member is interested in serving in ministries',
      field: 'interested_in_serving'
    }
  }, {
    sequelize,
    modelName: 'Member',
    tableName: 'members',
    timestamps: true,
    underscored: true
  });

  return Member;
}; 

File: backendJava/src/main/java/church/abunearegawi/backend/model/Member.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "members")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Member {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "firebase_uid", unique = true)
    private String firebaseUid;

    @Column(name = "first_name", nullable = false)
    private String firstName;

    @Column(name = "middle_name")
    private String middleName;

    @Column(name = "last_name", nullable = false)
    private String lastName;

    @Column(unique = true)
    private String email;

    @Column(name = "phone_number", nullable = false, unique = true)
    private String phoneNumber;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Enumerated(EnumType.STRING)
    private Gender gender;

    // marital_status and interested_in_serving are Postgres named enum types
    // (enum_members_marital_status, enum_members_interested_in_serving), not
    // varchar. @Enumerated(EnumType.STRING) alone binds a varchar, and Postgres
    // will not implicitly cast it, so every member INSERT failed with
    // "column is of type enum_... but expression is of type character varying".
    // NAMED_ENUM binds the value as the enum type. Any entity mapping one of the
    // schema's 23 enum columns needs the same treatment.
    @Column(name = "marital_status")
    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    private MaritalStatus maritalStatus;

    @Column(name = "baptism_name")
    private String baptismName;

    @Column(name = "repentance_father")
    private String repentanceFather;

    @Column(name = "household_size", nullable = false)
    @Builder.Default
    private Integer householdSize = 1;

    @Column(name = "street_line1")
    private String streetLine1;

    @Column(name = "apartment_no")
    private String apartmentNo;

    private String city;
    private String state;

    @Column(name = "postal_code")
    private String postalCode;

    @Builder.Default
    private String country = "USA";

    @Column(name = "emergency_contact_name")
    private String emergencyContactName;

    @Column(name = "emergency_contact_phone")
    private String emergencyContactPhone;

    @Column(name = "yearly_pledge", precision = 10, scale = 2)
    private BigDecimal yearlyPledge;

    @Column(name = "date_joined_parish")
    private LocalDate dateJoinedParish;

    @Column(name = "spouse_name")
    private String spouseName;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "family_id")
    private Member familyHead;

    @OneToMany(mappedBy = "familyHead")
    private List<Member> familyMembers;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private Role role = Role.member;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    @Builder.Default
    private String roles = "[]";

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private boolean isActive = true;

    @Column(name = "is_imported", nullable = false)
    @Builder.Default
    private boolean isImported = false;

    @Column(name = "registration_status", nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private RegistrationStatus registrationStatus = RegistrationStatus.pending;

    @Column(name = "is_welcomed", nullable = false)
    @Builder.Default
    private boolean isWelcomed = false;

    @Column(name = "welcomed_at")
    private LocalDateTime welcomedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "welcomed_by")
    private Member welcomedBy;

    @Column(name = "interested_in_serving")
    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Builder.Default
    private InterestedInServing interestedInServing = InterestedInServing.maybe;

    // isBaptized, medicalConditions, allergies, medications, dietaryRestrictions
    // and notes were declared here but belong to Dependent, which is where the
    // Node model defines them and where the columns actually exist. Hibernate
    // names every column in its SELECT, so their presence here made the first
    // read of any member fail with "column m1_0.allergies does not exist".
    // See Dependent for the real fields; see MemberRepositorySchemaTest.

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // Enums
    public enum Gender {
        male, female, other, Male, Female
    }

    public enum MaritalStatus {
        single, married, divorced, widowed, Single, Married, Divorced, Widowed
    }

    public enum Role {
        member, admin, treasurer, secretary, church_leadership, relationship, guest, deacon, priest, dependent,
        bookkeeper, budget_committee, auditor, ar_team, ap_team,
        Member, Admin, Treasurer, Secretary, Church_Leadership, Relationship, Guest, Deacon, Priest, Dependent,
        Bookkeeper, Budget_Committee, Auditor, AR_Team, AP_Team
    }

    public enum RegistrationStatus {
        pending, complete, incomplete
    }

    public enum InterestedInServing {
        yes, no, maybe
    }
}
